File: CUDC/CUDC.Web/Services/AdminService.cs
using CUDC.Common.Dtos;
using CUDC.Common.Dtos.UserSearch;
using Microsoft.Extensions.Configuration;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Void = CUDC.Common.Util.Void;

namespace CUDC.Web.Services
{
    public class AdminService : ServiceBase, IAdminService
    {
        private readonly IConfiguration _config;

        public AdminService(IConfiguration config) : base(config)
        {
            _config = config;
        }

        public async Task<IEnumerable<SurveyDto>> GetSurveys()
        {
            var requestUri = WebApiUrl("Admin/GetSurveys");
            return await GetAsync<List<SurveyDto>>(requestUri);
        }

        public async Task<SurveyDto> GetSurvey(Guid surveyId)
        {
            var requestUri = WebApiUrl($"Admin/GetSurvey/{surveyId}");
            return await GetAsync<SurveyDto>(requestUri);
        }

        public async Task CreateSurvey(SurveyDto survey)
        {
            var requestUri = WebApiUrl("Admin/CreateSurvey");
            await PostAsync<Void>(requestUri, survey);
        }

        public async Task UpdateSurvey(SurveyDto survey)
        {
            var requestUri = WebApiUrl("Admin/UpdateSurvey");
            await PostAsync<Void>(requestUri, survey);
        }

        public async Task CopySurvey(CopySurveyRequest request)
        {
            var requestUri = WebApiUrl($"Admin/CopySurvey");
            await PostAsync<Void>(requestUri, request);
        }

        public async Task CreateQuestions(QuestionsEdit questions)
        {
            var requestUri = WebApiUrl("Admin/CreateQuestions");
            await PostAsync<Void>(requestUri, questions);
        }

        public async Task UpdateQuestions(QuestionsEdit questions)
        {
            var requestUri = WebApiUrl("Admin/UpdateQuestions");
            await PostAsync<Void>(requestUri, questions);
        }

        public async Task<IEnumerable<ResponseDto>> GetResponses(Guid surveyId)
        {
            var requestUri = WebApiUrl($"Admin/GetResponses/{surveyId}");
            return await GetAsync<List<ResponseDto>>(requestUri);
        }

        public async Task RejectResponse(Guid responseId)
        {
            var requestUri = WebApiUrl($"Admin/RejectResponse/{responseId}");
            await GetAsync<Void>(requestUri);
        }

        public async Task<IEnumerable<UserRoleDto>> GetUsers()
        {
            var requestUri = WebApiUrl("Admin/GetUsers");
            return await GetAsync<List<UserRoleDto>>(requestUri);
        }

        public async Task CreateUser(UserRoleDto user)
        {
            var requestUri = WebApiUrl("Admin/CreateUser");
            await PostAsync<Void>(requestUri, user);
        }

        public async Task UpdateUser(UserRoleDto user)
        {
            var requestUri = WebApiUrl("Admin/UpdateUser");
            await PostAsync<Void>(requestUri, user);
        }

        public async Task DeleteUser(Guid id)
        {
            var requestUri = WebApiUrl($"Admin/DeleteUser/{id}");
            await GetAsync<Void>(requestUri);
        }

        private string WebApiUrl(string path)
        {
            return $"{_config["WebApiBaseUrl"]}{path}";
        }

        public async Task<IEnumerable<CycleDateDto>> GetCycleDates()
        {
            var requestUri = WebApiUrl("Admin/GetCycleDates");
            return await GetAsync<List<CycleDateDto>>(requestUri);
        }

        public async Task DeleteSurvey(Guid id)
        {
            var requestUri = WebApiUrl($"Admin/DeleteSurvey/{id}");
            await GetAsync<Void>(requestUri);
        }

        public async Task<IEnumerable<SearchUserResult>> GetSearchUserResults(SearchUserRequest request)
        {
            var requestUri = WebApiUrl("Admin/GetSearchUserResults");
            return await PostAsync<List<SearchUserResult>>(requestUri, request);
        }

    }
}

File: CUDC/CUDC.Client/src/app/services/admin.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { CycleDate } from '../models/auth/cycle-date';
import { UserRole } from '../models/auth/user-role';
import { SearchUserRequest } from '../models/search/search-user-request';
import { SearchUserResult } from '../models/search/search-user-result';
import { QuestionsEdit } from '../models/survey/questions-edit';
import { Response } from '../models/survey/response';
import { Survey } from '../models/survey/survey';
import { ApiService } from './api.service';

@Injectable({
  providedIn: 'root'
})
export class AdminService extends ApiService {
  constructor(httpClient: HttpClient) {
    super(httpClient);
  }

  getSurveys(): Observable<Survey[]> {
    return this.httpClient.get<Survey[]>(this.url('Admin/GetSurveys'));
  }

  getSurvey(surveyId: string): Observable<Survey> {
    return this.httpClient.get<Survey>(this.url(`Admin/GetSurvey/${surveyId}`));
  }

  createSurvey(survey: Survey): Observable<Survey> {
    return this.httpClient.post<Survey>(this.url('Admin/CreateSurvey'), survey);
  }

  updateSurvey(survey: Survey): Observable<Survey> {
    return this.httpClient.put<Survey>(this.url('Admin/UpdateSurvey'), survey);
  }

  copySurvey(surveyId: string) {
    return this.httpClient.get(this.url(`Admin/CopySurvey/${surveyId}`));
  }

  getQuestions(surveyId: string): Observable<QuestionsEdit> {
    return this.httpClient.get<QuestionsEdit>(this.url(`Admin/GetQuestions/${surveyId}`));
  }

  // createQuestions(questions: QuestionsEdit) {
  //   return this.httpClient.post(this.url('Admin/CreateQuestions'), questions);
  // }

  updateQuestions(questions: QuestionsEdit) {
    return this.httpClient.put(this.url('Admin/UpdateQuestions'), questions);
  }

  getResponses(surveyId: string): Observable<Response[]> {
    return this.httpClient.get<Response[]>(this.url(`Admin/GetResponses/${surveyId}`));
  }

  rejectResponse(responseId: string) {
    return this.httpClient.patch(this.url(`Admin/RejectResponse/${responseId}`), {});
  }

  getUsers(): Observable<UserRole[]> {
    return this.httpClient.get<UserRole[]>(this.url('Admin/GetUsers'));
  }

  createUser(user: UserRole): Observable<UserRole> {
    return this.httpClient.post<UserRole>(this.url('Admin/CreateUser'), user);
  }

  updateUser(user: UserRole): Observable<UserRole> {
    return this.httpClient.put<UserRole>(this.url('Admin/UpdateUser'), user);
  }

  deleteUser(id: string) {
    return this.httpClient.delete(this.url(`Admin/DeleteUser/${id}`));
  }

  getCycleDates(): Observable<CycleDate[]> {
    return this.httpClient.get<CycleDate[]>(this.url('Admin/GetCycleDates'));
  }  

  deleteSurvey(surveyId: string) {
    return this.httpClient.get(this.url(`Admin/DeleteSurvey/${surveyId}`));
  }

  
  getSearchUserResults(request: SearchUserRequest): Observable<SearchUserResult[]> {
    return this.httpClient.post<SearchUserResult[]>(this.url('Admin/GetSearchUserResults'), request);
  }
}
